Parsimonious stubs

jpy-git opened the pull request:
[Parsimonious](https://github.com/erikrose/parsimonious) is a PEG parser package.

PyPI link: https://pypi.org/project/parsimonious/

I'm making use of this package in a couple projects and there's also an open [issue](https://github.com/erikrose/parsimonious/issues/183) about adding type hints so figured I'd pick this up.

JelleZijlstra commented on stubs/parsimonious/parsimonious/exceptions.pyi:
```diff
@@ -0,0 +1,24 @@
+from parsimonious.expressions import Expression as Expression
```
Most of these (outside `__init__.pyi`) probably shouldn't be using `as`; that indicates that the file is re-exporting the name.

JelleZijlstra commented on stubs/parsimonious/parsimonious/__init__.pyi:
```diff
@@ -0,0 +1,8 @@
+from parsimonious.exceptions import (
```
I'm also seeing `VisitationError` at runtime

JelleZijlstra commented on stubs/parsimonious/parsimonious/expressions.pyi:
```diff
@@ -0,0 +1,68 @@
+_CALLABLE_RETURN_TYPE = int | tuple[int, list[Node]] | Node | None
+_CALLABLE_TYPE = (
+    Callable[[str, int], _CALLABLE_RETURN_TYPE]
+    | Callable[[str, int, dict[tuple[int, int], Node], ParseError, Grammar], _CALLABLE_RETURN_TYPE]
+    | Callable[[Any, str, int], _CALLABLE_RETURN_TYPE]
```
At runtime `expression()` raises an error if it doesn't take either 2 or 5 arguments, so do we need the 3-arg and 6-arg variants?

jpy-git replied:
So these account for this case: https://github.com/erikrose/parsimonious/blob/3da7e804c07d4e495873be208701b5c955247c58/parsimonious/expressions.py#L72-L74

Essentially it's meant to cover when methods are supplied as the custom callable e.g. `callable=some_instance.some_method`. The code just assumes the first argument is `self` in that case and removes it and then checks the length is 2 or 5.

I put `Any` to represent this `self` param since it could be for any custom class. lmk if there's a better way to type it though 😄

jpy-git replied:
Actually apologies, just checked that code for the latest release of the package and these method arguments weren't available at that point so will remove them now: https://github.com/erikrose/parsimonious/blob/f4496cc0b6bae44c09eaaf4233e94a8d70e396a9/parsimonious/expressions.py#L60-L67

JelleZijlstra commented on stubs/parsimonious/parsimonious/expressions.pyi:
```diff
@@ -0,0 +1,68 @@
+
+class Compound(Expression):
+    members: typing.Sequence[Expression]
+    def __init__(self, *members: typing.Sequence[Expression], **kwargs: Any) -> None: ...
```
```suggestion
    def __init__(self, *members: Expression, **kwargs: Any) -> None: ...
```
The `*args` annotation is for the type of each arg, not the whole of args.

jpy-git replied:
ahh yep my bad!

JelleZijlstra commented on stubs/parsimonious/parsimonious/grammar.pyi:
```diff
@@ -0,0 +1,69 @@
+)
+from parsimonious.nodes import Node as Node, NodeVisitor as NodeVisitor
+from parsimonious.utils import evaluate_string as evaluate_string
+from six import text_type
```
Can just use `str`, we don't need to care about Python 2

JelleZijlstra commented on stubs/parsimonious/parsimonious/grammar.pyi:
```diff
@@ -0,0 +1,69 @@
+    name: str
+
+class RuleVisitor(NodeVisitor):
+    quantifier_classes: dict[str, Expression]
```
```suggestion
    quantifier_classes: dict[str, type[Expression]]
```

JelleZijlstra commented on stubs/parsimonious/parsimonious/grammar.pyi:
```diff
@@ -0,0 +1,69 @@
+    visit_expression: Callable[[Self, Node, list[Any]], Any]
+    visit_term: Callable[[Self, Node, list[Any]], Any]
+    visit_atom: Callable[[Self, Node, list[Any]], Any]
+    custom_rules: dict[str, Expression]
```
Also maybe should have `type[]`?

jpy-git replied:
So this is used to updated the rule_map which contains instantiated Expressions so think this is correct as is?

https://github.com/erikrose/parsimonious/blob/3da7e804c07d4e495873be208701b5c955247c58/parsimonious/grammar.py#L438-L447

jpy-git commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+class RegexNode(Node):
+    match: Match[str]
+
+class RuleDecoratorMeta(type):
+    def __new__(metaclass: type[Self], name: str, bases: tuple[type, ...], namespace: dict[str, Any]) -> Self: ...
```
Pyright is complaining about this line:
`error: __new__ override should take a "cls" parameter (reportSelfClsParameterName)`

The error makes sense, but this is how it's written in the source code. `metaclass` gets passed to the `cls` parameter of `type.__new__`.

I tried adding this method to the allowlist but doesn't seem to have suppressed the error?

JelleZijlstra replied:
That allowlist is for stubtest, a different tool than pyright. To shut up pyright here you'd have to use `# type: ignore`.

jpy-git replied:
Perfect, good to know!

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+    start: int
+    end: int
+    children: list[Node]
+    def __init__(self, expr: Expression, full_text: str, start: int, end: int, children: list[Node] | None = ...) -> None: ...
```
Does it have to be a list? Would be preferable to write `Sequence` so we don't run into list invariance.

jpy-git replied:
Makes sense, will update the various visit methods similarly as well

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+from re import Match
+from typing import Any, Callable, Iterator, NoReturn, TypeVar
+from typing_extensions import Self
```
We can't use PEP 673 yet because mypy doesn't support it

jpy-git replied:
good shout, I've removed the function it was used in anyway 👍

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+    match: Match[str]
+
+class RuleDecoratorMeta(type):
+    def __new__(metaclass: type[Self], name: str, bases: tuple[type, ...], namespace: dict[str, Any]) -> Self: ...
```
I think we can omit this method, since the signature is the same as the base class.

jpy-git replied:
fine by me thanks 👍

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+class RuleDecoratorMeta(type):
+    def __new__(metaclass: type[Self], name: str, bases: tuple[type, ...], namespace: dict[str, Any]) -> Self: ...
+
+class NodeVisitor(type):
```
```suggestion
class NodeVisitor(metaclass=RuleDecoratorMeta):
```

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+
+class NodeVisitor(type):
+    grammar: Grammar | Any
+    unwrapped_exceptions: tuple[Exception]
```
```suggestion
    unwrapped_exceptions: tuple[type[Exception], ...]
```
`tuple[T]` means a one-tuple. And I suspect this holds exception classes, not instances.

jpy-git replied:
You're correct thanks!

JelleZijlstra commented on stubs/parsimonious/parsimonious/nodes.pyi:
```diff
@@ -0,0 +1,40 @@
+    def match(self, text: str, pos: int = ...) -> Node: ...
+    def lift_child(self, node: Node, children: list[Any]) -> Any: ...
+
+_T = TypeVar("_T")
+
+def rule(rule_string: str) -> Callable[[Callable[..., _T]], Callable[..., _T]]: ...
```
```suggestion
_CallableT = TypeVar("_CallableT", bound=Callable[..., Any])

def rule(rule_string: str) -> Callable[[_CallableT], _CallableT]: ...
```

jpy-git commented:
@JelleZijlstra Thanks for the thorough review 😄 

I believe I've addressed the comments. I'm having some trouble with getting [Check third party stubs with stubtest](https://github.com/python/typeshed/runs/5525076052?check_suite_focus=true) to pass since it's complaining about some of the subclass types not agreeing with the superclass types, which is what's happening in the code, and stubtest_allowlist doesn't seem to fix things. Any ideas what I'm missing?

hauntsaninja commented:
stubtest runs a mypy build first to understand what's in the stubs. The mypy build is failing, so we haven't even gotten to the actual stubtest logic, which is what stubtest allowlists affect.

Assuming your subclass types are actually correct, adding type ignore will allow the build to proceed. For the "Type application has too many types", it's probably some mypy bug, try using an old style typing.Union + typing.Tuple

JelleZijlstra commented:
The errors are from mypy itself, not stubtest; they'll need `# type: ignore`. I'll try to figure out why the regular mypy test didn't catch this.

JelleZijlstra commented:
@hauntsaninja it's not expected that the mypy test in CI is passing though, right? Looks like it runs only on the stdlib.

hauntsaninja commented:
Yeah, looks like it's only running against the stdlib. It's not the first time I've seen stubtest third party catch mypy errors that mypy hasn't caught, so the test has probably been buggy for a while.

JelleZijlstra commented:
It's likely related to the mypy feature where it does `os._exit` to exit faster.

JelleZijlstra commented:
Fix in #7478
